feat: Added enclave pool for improving performance on enclave creation

leoporoli opened the pull request:
## Description:
- Added enclave pool for improving performance on enclave creation
- The `pool-size` new flag was added in the `engine start and engine restart` commands

## What is pending? 
- reuse idle enclaves when the engine is restarted

## Is this change user-facing?
YES

## References (if applicable):
[Docs are here](https://www.notion.so/kurtosistech/Enclave-Pool-Project-7273552e53e44b0f931a57568cd170a0?d=2f870450d01d4ad7b8af6d0045602a2a#a20862960dad4868b0b9b6eb320b6178)

leoporoli committed:
First version before moving enclave pool inside the enclave manager

leoporoli committed:
Added enclave pool inside the enclave manager and moved the creation logic inside the enclave creator

leoporoli committed:
Added private method to evaluate if it's allowed or not

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
Implementing reuse or remove previous idle enclaves

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
Added remove old idle enclaves

leoporoli committed:
Moved enclave name creation logic to the enclave manager

leoporoli committed:
Added pool-size into Kurtosis CLI

leoporoli committed:
Docs updated

leoporoli committed:
Removed unused code and improving comments

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
Change for passing S1002 linter check

gbouv commented on docs/docs/cli-reference/engine-start.md:
```diff
@@ -13,4 +13,5 @@ This command will do nothing if the Kurtosis engine is already running.
-* `--version`: The version (Docker tag) of the Kurtosis engine that should be started. If not set, the engine will start up with the default version.
\ No newline at end of file
+* `--version`: The version (Docker tag) of the Kurtosis engine that should be started. If not set, the engine will start up with the default version.
+* `--pool-size`: The size of the Kurtosis engine enclave pool. The enclave pool is a component of the Kurtosis engine that allows us to create and maintain 'n' number of idle enclaves for future use. This functionality allows to improve the performance for each new creation enclave request.
```
Should we add that this is only available for Kubernetes backend? BTW what happens if I do `kurtosis engine start --pool-size 2` while using a Docker backend?

leoporoli replied:
good catch! I added the CAUTION on the docs and a validation on the CLI

gbouv commented on engine/server/engine/enclave_manager/enclave_pool.go:
```diff
@@ -0,0 +1,314 @@
+		return nil, stacktrace.Propagate(err, "An error occurred destroying previous idle enclave before creating the enclave pool")
+	}
+
+	go enclavePool.run()
```
Will that automatically be killed if we stop the engine? Or do we need a close function of some sort we can call when stopping the engine?
This would also have the advantage of killing the idle enclaves when we stop the engine, not when we start a new one

leoporoli replied:
yes, it's pending to implement a `StopEngine` engine endpoint for being used when users execute `engine stop` and the killing idle enclave process in the start if for when the engine crashes and engine restarts (it's pending to implement the reuse idle enclave mechanism also)

gbouv commented on engine/server/engine/enclave_manager/enclave_pool.go:
```diff
@@ -0,0 +1,314 @@
+		}
+		enclaveUUID := enclave.EnclaveUUID(newEnclaveInfo.EnclaveUuid)
+
+		pool.idleEnclavesChan <- enclaveUUID
```
So, you start by creating the enclave, and THEN you push it to the channel?? Does this work in practice? Won't you end up with one more enclave running than the desired size of the pool?

gbouv replied:
Ah yes that's the comment you wrote in CreateEnclavePool function. I understand that the enclave UUID can be retrieved only AFTER the enclave is created, therefore you have to create the enclave first and only after push it to the channel. Can't we use a distinct channel for the "pooling" mechanism, and store the enclave UUIDs in a simple array for example?
IDK, this seems like an extra complexity here

leoporoli replied:
@gbouv I'm not really sure if I understand your concern. What error do you think that could happen with the current state?

leoporoli replied:
it was refactored, now the channel length is equal to the pool size

gbouv commented on engine/server/engine/enclave_manager/enclave_pool.go:
```diff
@@ -0,0 +1,314 @@
+
+// The enclave pool feature is only available for Kubernetes so far, and it will be activated
+// only if users require this when setting the pool-size value
+func isEnclavePoolAllowedForThisConfig(
```
Where is this used?

leoporoli replied:
@gbouv It's used in the `EnclaveManager` here:
<img width="865" alt="image" src="https://github.com/kurtosis-tech/kurtosis/assets/29951623/97cfcfac-380e-405e-90dd-360a2fb4864b">
but now we a second though I think that I'll move these validations inside the `CreateEnclavePool` method and I'll add the validations in the CLI also

leoporoli replied:
it was refactored as I described in the previous comment

gbouv reviewed (approved):
Looks generally good, not a huge fan of the "create enclave first and then push it to the channel" logic though, I'd prefer the "push to channel" and "enclave creation" operation to happen as closely as possible

leoporoli committed:
Added backend type validation and renamed the flag

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
Fix

Commit:
fix: Fix TS proto bindings (#797)

## Description:


## Is this change user facing?
NO
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful
information. -->

Commit:
build: Re-enable typescript integration testsuites (#799)

## Description:
Re-enable typescript test suites. I am not entirely sure why they were
disabled, though they were completely broken and I had to make a bunch
of changes to fix them. Maybe that was the actual reason, not sure

## Is this change user facing?
NO
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful
information. -->

dependabot[bot] committed:
build(deps): Bump semver from 7.3.7 to 7.5.2 in /api/typescript (#772)

Bumps [semver](https://github.com/npm/node-semver) from 7.3.7 to 7.5.2.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/npm/node-semver/releases">semver's
releases</a>.</em></p>
<blockquote>
<h2>v7.5.2</h2>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.5.1...v7.5.2">7.5.2</a>
(2023-06-15)</h2>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/58c791f40ba8cf4be35a5ca6644353ecd6249edc"><code>58c791f</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/566">#566</a>
diff when detecting major change from prerelease (<a
href="https://redirect.github.com/npm/node-semver/issues/566">#566</a>)
(<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/5c8efbcb3c6c125af10746d054faff13e8c33fbd"><code>5c8efbc</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/565">#565</a>
preserve build in raw after inc (<a
href="https://redirect.github.com/npm/node-semver/issues/565">#565</a>)
(<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/717534ee353682f3bcf33e60a8af4292626d4441"><code>717534e</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/564">#564</a>
better handling of whitespace (<a
href="https://redirect.github.com/npm/node-semver/issues/564">#564</a>)
(<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
</ul>
<h2>v7.5.1</h2>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.5.0...v7.5.1">7.5.1</a>
(2023-05-12)</h2>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/d30d25a5c1fb963c3cc9178cb1769fe45e4a3cab"><code>d30d25a</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/559">#559</a>
show type on invalid semver error (<a
href="https://redirect.github.com/npm/node-semver/issues/559">#559</a>)
(<a
href="https://github.com/tjenkinson"><code>@​tjenkinson</code></a>)</li>
</ul>
<h2>v7.5.0</h2>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.4.0...v7.5.0">7.5.0</a>
(2023-04-17)</h2>
<h3>Features</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/503a4e52fe2b1c6ed1400d33149f7733c8361eed"><code>503a4e5</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/548">#548</a>
allow identifierBase to be false (<a
href="https://redirect.github.com/npm/node-semver/issues/548">#548</a>)
(<a href="https://github.com/lsvalina"><code>@​lsvalina</code></a>)</li>
</ul>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/e219bb454036a0c23e34407591f921c8edb688e7"><code>e219bb4</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/552">#552</a>
throw on bad version with correct error message (<a
href="https://redirect.github.com/npm/node-semver/issues/552">#552</a>)
(<a
href="https://github.com/wraithgar"><code>@​wraithgar</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/fc2f3df0b5d25253b3580607e111a9a280d888ca"><code>fc2f3df</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/546">#546</a>
incorrect results from diff sometimes with prerelease versions (<a
href="https://redirect.github.com/npm/node-semver/issues/546">#546</a>)
(<a
href="https://github.com/tjenkinson"><code>@​tjenkinson</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/27817677794f592b592bf6181a80a4824ff762b2"><code>2781767</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/547">#547</a>
avoid re-instantiating SemVer during diff compare (<a
href="https://redirect.github.com/npm/node-semver/issues/547">#547</a>)
(<a href="https://github.com/macno"><code>@​macno</code></a>)</li>
</ul>
<h2>v7.4.0</h2>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.3.8...v7.4.0">7.4.0</a>
(2023-04-10)</h2>
<h3>Features</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/113f51312a1a6b6aa50d4f9486b4fde21782c1f5"><code>113f513</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/532">#532</a>
identifierBase parameter for .inc (<a
href="https://redirect.github.com/npm/node-semver/issues/532">#532</a>)
(<a href="https://github.com/wraithgar"><code>@​wraithgar</code></a>, <a
href="https://github.com/b-bly"><code>@​b-bly</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/48d8f8fa63bf6e35db70ff840b6da1a51596a5a8"><code>48d8f8f</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/530">#530</a>
export new RELEASE_TYPES constant (<a
href="https://github.com/hcharley"><code>@​hcharley</code></a>)</li>
</ul>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/940723d22bca824993627c45ac30dd3d2854b8cd"><code>940723d</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/538">#538</a>
intersects with v0.0.0 and v0.0.0-0 (<a
href="https://redirect.github.com/npm/node-semver/issues/538">#538</a>)
(<a
href="https://github.com/wraithgar"><code>@​wraithgar</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/aa516b50b32f5a144017d8fc1b9efe0540963c91"><code>aa516b5</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/535">#535</a>
faster parse options (<a
href="https://redirect.github.com/npm/node-semver/issues/535">#535</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/61e6ea1e9b7af01baf19ab0c0a63c8e3ebfac97c"><code>61e6ea1</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/536">#536</a>
faster cache key factory for range (<a
href="https://redirect.github.com/npm/node-semver/issues/536">#536</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/f8b8b619e71746a47852a9d301f3087ab311444f"><code>f8b8b61</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/541">#541</a>
optimistic parse (<a
href="https://redirect.github.com/npm/node-semver/issues/541">#541</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/796cbe29b06d102e1b16f3ed78eaba210ece951e"><code>796cbe2</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/533">#533</a>
semver.diff prerelease to release recognition (<a
href="https://redirect.github.com/npm/node-semver/issues/533">#533</a>)
(<a href="https://github.com/wraithgar"><code>@​wraithgar</code></a>, <a
href="https://github.com/dominique-blockchain"><code>@​dominique-blockchain</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/3f222b144033525ca9f8a2ce5bc6e02f0401881f"><code>3f222b1</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/537">#537</a>
reuse comparators on subset (<a
href="https://redirect.github.com/npm/node-semver/issues/537">#537</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/f66cc45c6e82eebb4b5b51af73e7b8dcaeda7e21"><code>f66cc45</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/539">#539</a>
faster diff (<a
href="https://redirect.github.com/npm/node-semver/issues/539">#539</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
</ul>
<h3>Documentation</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/c5d29df6f75741fea27fffe3b88c9c3b28e3ca73"><code>c5d29df</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/530">#530</a>
Add &quot;Constants&quot; section to README (<a
href="https://github.com/hcharley"><code>@​hcharley</code></a>)</li>
</ul>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a
href="https://github.com/npm/node-semver/blob/main/CHANGELOG.md">semver's
changelog</a>.</em></p>
<blockquote>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.5.1...v7.5.2">7.5.2</a>
(2023-06-15)</h2>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/58c791f40ba8cf4be35a5ca6644353ecd6249edc"><code>58c791f</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/566">#566</a>
diff when detecting major change from prerelease (<a
href="https://redirect.github.com/npm/node-semver/issues/566">#566</a>)
(<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/5c8efbcb3c6c125af10746d054faff13e8c33fbd"><code>5c8efbc</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/565">#565</a>
preserve build in raw after inc (<a
href="https://redirect.github.com/npm/node-semver/issues/565">#565</a>)
(<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/717534ee353682f3bcf33e60a8af4292626d4441"><code>717534e</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/564">#564</a>
better handling of whitespace (<a
href="https://redirect.github.com/npm/node-semver/issues/564">#564</a>)
(<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
</ul>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.5.0...v7.5.1">7.5.1</a>
(2023-05-12)</h2>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/d30d25a5c1fb963c3cc9178cb1769fe45e4a3cab"><code>d30d25a</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/559">#559</a>
show type on invalid semver error (<a
href="https://redirect.github.com/npm/node-semver/issues/559">#559</a>)
(<a
href="https://github.com/tjenkinson"><code>@​tjenkinson</code></a>)</li>
</ul>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.4.0...v7.5.0">7.5.0</a>
(2023-04-17)</h2>
<h3>Features</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/503a4e52fe2b1c6ed1400d33149f7733c8361eed"><code>503a4e5</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/548">#548</a>
allow identifierBase to be false (<a
href="https://redirect.github.com/npm/node-semver/issues/548">#548</a>)
(<a href="https://github.com/lsvalina"><code>@​lsvalina</code></a>)</li>
</ul>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/e219bb454036a0c23e34407591f921c8edb688e7"><code>e219bb4</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/552">#552</a>
throw on bad version with correct error message (<a
href="https://redirect.github.com/npm/node-semver/issues/552">#552</a>)
(<a
href="https://github.com/wraithgar"><code>@​wraithgar</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/fc2f3df0b5d25253b3580607e111a9a280d888ca"><code>fc2f3df</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/546">#546</a>
incorrect results from diff sometimes with prerelease versions (<a
href="https://redirect.github.com/npm/node-semver/issues/546">#546</a>)
(<a
href="https://github.com/tjenkinson"><code>@​tjenkinson</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/27817677794f592b592bf6181a80a4824ff762b2"><code>2781767</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/547">#547</a>
avoid re-instantiating SemVer during diff compare (<a
href="https://redirect.github.com/npm/node-semver/issues/547">#547</a>)
(<a href="https://github.com/macno"><code>@​macno</code></a>)</li>
</ul>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.3.8...v7.4.0">7.4.0</a>
(2023-04-10)</h2>
<h3>Features</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/113f51312a1a6b6aa50d4f9486b4fde21782c1f5"><code>113f513</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/532">#532</a>
identifierBase parameter for .inc (<a
href="https://redirect.github.com/npm/node-semver/issues/532">#532</a>)
(<a href="https://github.com/wraithgar"><code>@​wraithgar</code></a>, <a
href="https://github.com/b-bly"><code>@​b-bly</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/48d8f8fa63bf6e35db70ff840b6da1a51596a5a8"><code>48d8f8f</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/530">#530</a>
export new RELEASE_TYPES constant (<a
href="https://github.com/hcharley"><code>@​hcharley</code></a>)</li>
</ul>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/940723d22bca824993627c45ac30dd3d2854b8cd"><code>940723d</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/538">#538</a>
intersects with v0.0.0 and v0.0.0-0 (<a
href="https://redirect.github.com/npm/node-semver/issues/538">#538</a>)
(<a
href="https://github.com/wraithgar"><code>@​wraithgar</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/aa516b50b32f5a144017d8fc1b9efe0540963c91"><code>aa516b5</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/535">#535</a>
faster parse options (<a
href="https://redirect.github.com/npm/node-semver/issues/535">#535</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/61e6ea1e9b7af01baf19ab0c0a63c8e3ebfac97c"><code>61e6ea1</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/536">#536</a>
faster cache key factory for range (<a
href="https://redirect.github.com/npm/node-semver/issues/536">#536</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/f8b8b619e71746a47852a9d301f3087ab311444f"><code>f8b8b61</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/541">#541</a>
optimistic parse (<a
href="https://redirect.github.com/npm/node-semver/issues/541">#541</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/796cbe29b06d102e1b16f3ed78eaba210ece951e"><code>796cbe2</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/533">#533</a>
semver.diff prerelease to release recognition (<a
href="https://redirect.github.com/npm/node-semver/issues/533">#533</a>)
(<a href="https://github.com/wraithgar"><code>@​wraithgar</code></a>, <a
href="https://github.com/dominique-blockchain"><code>@​dominique-blockchain</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/3f222b144033525ca9f8a2ce5bc6e02f0401881f"><code>3f222b1</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/537">#537</a>
reuse comparators on subset (<a
href="https://redirect.github.com/npm/node-semver/issues/537">#537</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/f66cc45c6e82eebb4b5b51af73e7b8dcaeda7e21"><code>f66cc45</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/539">#539</a>
faster diff (<a
href="https://redirect.github.com/npm/node-semver/issues/539">#539</a>)
(<a href="https://github.com/H4ad"><code>@​H4ad</code></a>)</li>
</ul>
<h3>Documentation</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/c5d29df6f75741fea27fffe3b88c9c3b28e3ca73"><code>c5d29df</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/530">#530</a>
Add &quot;Constants&quot; section to README (<a
href="https://github.com/hcharley"><code>@​hcharley</code></a>)</li>
</ul>
<h2><a
href="https://github.com/npm/node-semver/compare/v7.3.7...v7.3.8">7.3.8</a>
(2022-10-04)</h2>
<h3>Bug Fixes</h3>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/e7b78de06eb14a7fa2075cedf9f167040d8d31af"><code>e7b78de</code></a>
chore: release 7.5.2</li>
<li><a
href="https://github.com/npm/node-semver/commit/58c791f40ba8cf4be35a5ca6644353ecd6249edc"><code>58c791f</code></a>
fix: diff when detecting major change from prerelease (<a
href="https://redirect.github.com/npm/node-semver/issues/566">#566</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/5c8efbcb3c6c125af10746d054faff13e8c33fbd"><code>5c8efbc</code></a>
fix: preserve build in raw after inc (<a
href="https://redirect.github.com/npm/node-semver/issues/565">#565</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/717534ee353682f3bcf33e60a8af4292626d4441"><code>717534e</code></a>
fix: better handling of whitespace (<a
href="https://redirect.github.com/npm/node-semver/issues/564">#564</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/2f738e9a70d9b9468b7b69e9ed3e12418725c650"><code>2f738e9</code></a>
chore: bump <code>@​npmcli/template-oss</code> from 4.14.1 to 4.15.1 (<a
href="https://redirect.github.com/npm/node-semver/issues/558">#558</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/aa016a67162c195938f7873ea29a73dac47ff9ba"><code>aa016a6</code></a>
chore: release 7.5.1</li>
<li><a
href="https://github.com/npm/node-semver/commit/d30d25a5c1fb963c3cc9178cb1769fe45e4a3cab"><code>d30d25a</code></a>
fix: show type on invalid semver error (<a
href="https://redirect.github.com/npm/node-semver/issues/559">#559</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/09c69e23cdf6c69c51f83635482fff89ab2574e3"><code>09c69e2</code></a>
chore: bump <code>@​npmcli/template-oss</code> from 4.13.0 to 4.14.1 (<a
href="https://redirect.github.com/npm/node-semver/issues/555">#555</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/5b02ad7163a3ddcbcadf499e4f6195d6f2226dce"><code>5b02ad7</code></a>
chore: release 7.5.0</li>
<li><a
href="https://github.com/npm/node-semver/commit/e219bb454036a0c23e34407591f921c8edb688e7"><code>e219bb4</code></a>
fix: throw on bad version with correct error message (<a
href="https://redirect.github.com/npm/node-semver/issues/552">#552</a>)</li>
<li>Additional commits viewable in <a
href="https://github.com/npm/node-semver/compare/v7.3.7...v7.5.2">compare
view</a></li>
</ul>
</details>
<details>
<summary>Maintainer changes</summary>
<p>This version was pushed to npm by <a
href="https://www.npmjs.com/~npm-cli-ops">npm-cli-ops</a>, a new
releaser for semver since your current version.</p>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=semver&package-manager=npm_and_yarn&previous-version=7.3.7&new-version=7.5.2)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)
You can disable automated security fix PRs for this repo from the
[Security Alerts
page](https://github.com/kurtosis-tech/kurtosis/network/alerts).

</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Guillaume Bouvignies <guillaume.bouvignies@kurtosistech.com>

Commit:
refactor: Use streaming to download file via TS SDK (#793)

## Description:
Migrate to the streaming endpoint to download kurtosis enclave files via
the TS SDK.

## Is this change user facing?
NO
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful
information. -->

kurtosisbot committed:
chore(main): release 0.80.5 (#791)

:robot: I have created a release *beep* *boop*
---


##
[0.80.5](https://github.com/kurtosis-tech/kurtosis/compare/0.80.4...0.80.5)
(2023-06-30)


### Bug Fixes

* Fix TS proto bindings
([#797](https://github.com/kurtosis-tech/kurtosis/issues/797))
([7958dba](https://github.com/kurtosis-tech/kurtosis/commit/7958dba5cec3dfb09eb69f24785d33dbd94051d6))
* make dry run return the right return value
([#795](https://github.com/kurtosis-tech/kurtosis/issues/795))
([be5f6e7](https://github.com/kurtosis-tech/kurtosis/commit/be5f6e75229a3887dc84c7a139aebe84b09fc77d))
* More informative logging for instructions caching
([#785](https://github.com/kurtosis-tech/kurtosis/issues/785))
([376ac8c](https://github.com/kurtosis-tech/kurtosis/commit/376ac8ceb7085a744c5cf84756b5d2c72a2577f7))

---
This PR was generated with [Release
Please](https://github.com/googleapis/release-please). See
[documentation](https://github.com/googleapis/release-please#release-please).

---------

Co-authored-by: kurtosisbot <kurtosisbot@users.noreply.github.com>
Co-authored-by: Gyanendra Mishra <anomaly.the@gmail.com>
Co-authored-by: h4ck3rk3y <h4ck3rk3y@users.noreply.github.com>

Commit:
refactor: Use streaming to upload file via TS SDK (#796)

## Description:
Migrate to the streaming endpoint to upload kurtosis enclave files via
the TS SDK.

## Is this change user facing?
NO
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful
information. -->

Commit:
feat: Invert USE_INSTRUCTIONS_CACHING feature flag (#800)

## Description:
Replace USE_INSTRUCTIONS_CACHING feature flag with
NO_INSTRUCTIONS_CACHING. By default, instruction caching will happen

## Is this change user facing?
NO
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful
information. -->

thylocine33 committed:
docs: adds official k8s links to docs (#805)

## Description:
Added k8s official documents to https://docs.kurtosis.com/k8s

## Is this change user facing?
YES
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
#801

Co-authored-by: Karla Forester <karlaforester@Karlas-MacBook-Pro.local>

leoporoli committed:
chore: stop the enclave pool, which includes removing idle enclaves, when the engine is stopped (#798)

## Description:
stop the enclave pool, which includes removing idle enclaves, when the
engine is stopped

## Is this change user-facing?
NO

## References (if applicable):
[Docs are
here](https://www.notion.so/kurtosistech/Enclave-Pool-Project-7273552e53e44b0f931a57568cd170a0?d=2f870450d01d4ad7b8af6d0045602a2a#a20862960dad4868b0b9b6eb320b6178)

Related to #787

leoporoli committed:
Replaced rename enclave with update enclave

leoporoli committed:
Merge branch 'main' into lporoli/update-enclave

leoporoli committed:
Merge branch 'lporoli/update-enclave' into lporoli/enclave-pool

leoporoli committed:
Replaced rename enclave to update enclave

leoporoli committed:
Moved the Kurtosis backend type condition on top

leoporoli committed:
Moved the Kurtosis backend type condition before creating the enclave pool

leoporoli committed:
Fix destroyEnclavesByUUID

leoporoli committed:
comments

leoporoli committed:
Commented some jobs

leoporoli committed:
Rollbacked enclave pool to the original version

leoporoli committed:
Rollback CI config

leoporoli committed:
Rollbacked create engine

leoporoli committed:
Implement destroyIdleEnclavesFromPreviousRuns

leoporoli committed:
Moved validation on top

leoporoli committed:
chore: filter idle enclaves on `EnclaveManager` (#807)

## Description:
filter idle enclaves on `EnclaveManager` 
This work is part of the enclave pool project, and we need to filter
idle enclaves for instance when users get the enclaves from the engine
or when users call to the clean endpoint

## Is this change user-facing?
NO

## References (if applicable):

leoporoli committed:
Moved destroyIdleEnclavesFromPreviousRuns on top

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
Replace backend call with enclave names in memory for validate new enclave name

leoporoli committed:
Update enclave info instead of recreating it for avoiding Kubernetes backend calls

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
Removed unused vars

leoporoli committed:
Graceful engine and enclave pool stop

leoporoli committed:
Merge branch 'main' into lporoli/enclave-pool

leoporoli committed:
refactor CI test that checks containers after engine stop, now the container is removed when the engine is stopped
